### PR TITLE
LSTM conversion, training & multi-device support, and more

### LINKED ISSUE #8: Forward pass of Reshape
I'm having difficulty understanding what's going on in the forward pass of Reshape. It seems to me that `shape = [x if x != 0 else input.size(i) for i, x in enumerate(shape)]` is exactly what the ONNX spec calls for. Is the additional work being done for correctness purposes, or for speedup purposes?

Deleting the additional code is necessary to get onnx2pytorch to work on the Nvidia BERT-Large model [here](https://zenodo.org/record/3733868#.YImM6H1KilM), so perhaps there should be a flag set during __init__ that controls whether this additional work is performed.

### LINKED ISSUE #11: Handle indices output in MaxPool
If the ONNX MaxPool node has two outputs (ie Y and Indices), the corresponding nn.MaxPool{}d module should be initialized with return_indices=True.

(I hope to take care of this in my next PR...)

### LINKED ISSUE #17: NotImplementedError: Conversion not implemented for op_type=Loop.
please Help

### PR DESCRIPTION
**Pull request by calvinmccarter-at-lightmatter**
This PR contains a few key additions:

- Converts the ONNX LSTM operator
- Converts the ONNX Loop operator (resolves #17 )
- Supports training of converted models, by ensuring underlying arrays are copied from ONNX
- Makes model device-agnostic, by registering parameters/buffers and creating new tensors on the same device as inputs to each node
- Converts the ONNX BitShift, GatherND, NonMaxSuppression, PRelu, ReduceSum, Scatter, ScatterElements, ScatterND, Softplus, Softsign, ThresholdedRelu, Tile, and TopK operators
- Handles ops like MaxPool with multiple outputs (resolves #11)
- Tracks activation dependencies and deletes unneeded activations to save memory

This PR also contains features or bugfixes for the following ONNX nodes:

- Add (more aggressively reorders input for broadcasting)
- Div (because ONNX Div on int-types performs numpy.floor_divide which is different from torch.floor_divide)
- GlobalAveragePool (see https://github.com/onnx/onnx/pull/3472)
- Reshape (does not adjust shape if it's dynamically computed to account for changing input shape, resolves #8)
- Slice (handles negative steps, which is supported in ONNX and numpy, but not torch)
- Transpose (handles empty perm attribute)
- Where (gracefully handles mixed dtypes, which is ambiguous in the ONNX spec)

Finally, this PR contains more tests and minor formatting changes.

### CODE REVIEW

**Comment by Talmaj** [comment]
Thank you for your pull request! I'll have a look as soon as I can.

**Comment by calvinmccarter-at-lightmatter** [comment]
@Talmaj -- feel free to hold off reviewing this, because in a week or so I hope to add a few more features, and we may as well include it in this PR.

**Comment by calvinmccarter-at-lightmatter** [comment]
The tests are failing because tox doesn't detect the added requirement (torchvision) -- clearing the .tox cache should fix this. And I hope to finally finish adding stuff to this PR in the next day or so!

**Comment by calvinmccarter-at-lightmatter** [comment]
@Talmaj -- this PR is now finally ready for you to look at!

**Comment by Talmaj** [comment]
@calvinmccarter-at-lightmatter sorry for a late response, I was in holiday. I will have a look as soon as I can. Thanks!

**Comment by calvinmccarter-at-lightmatter** [comment]
@Talmaj , hope all is well -- do you think you'll be able to look at this soon?

File: onnx2pytorch/convert/model.py
Comment (Talmaj) [inline onnx2pytorch/convert/model.py:101]: Could you move this to a separate method? Call it: compute_activation_dependencies.
Reason is to keep functions and methods small and quickly understandable.

File: onnx2pytorch/convert/operations.py
Comment (Talmaj) [inline onnx2pytorch/convert/operations.py:33]: Wouldn't it make more sense to move Loop to operators/loop.py? Here doesn't seem like the right place.

File: onnx2pytorch/convert/operations.py
Comment (Talmaj) [inline onnx2pytorch/convert/operations.py:192]: Let's move the COMPOSITE_TYPES and others to `operations/__init__.py`
I would also rename them to:
COMPOSITE_LAYERS
STANDARD_LAYERS
MULTIOUTPUT_LAYERS
What do you think?

**Comment by Talmaj** [comment]
It's a large pull request, so I didn't really find time to finish the review yet. I've posted some comments for now. The loop test is failing for me. Check circleci.

**Comment by calvinmccarter-at-lightmatter** [comment]
The loop test seems to be broken starting with onnx==1.10.0, probably by https://github.com/onnx/onnx/pull/2918. I should probably make a PR to ONNX, but this at least fixes things on our end. I tested locally, and it now works with onnx=={1.8.0, 1.9.0, 1.10.0, and 1.10.1}.

File: onnx2pytorch/convert/operations.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/operations.py]: Done! (The imports in its `__init__` and `forward` are to avoid circular import errors.)

File: onnx2pytorch/convert/operations.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/operations.py]: Done! (I moved them to `constants.py` since these require external imports eg torch.nn)

File: onnx2pytorch/convert/layer.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/layer.py:34]: Yeah, so this is helpful is because of stuff we do when training the converted model. I can try condense it down into a simple test case, though, if you'd like.

File: onnx2pytorch/convert/model.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/model.py]: Done!

File: onnx2pytorch/operations/loop.py
Comment (Talmaj) [inline onnx2pytorch/operations/loop.py:19]: Loop seems to have a lot of similarities with ConvertModel, there could be some base class that both inherit, or one inherits the other one. For now no change is necessary, but let's open an issue with regards to this after the merge.

File: onnx2pytorch/operations/pad.py
Comment (Talmaj) [inline onnx2pytorch/operations/pad.py:16]: Why this change?

File: onnx2pytorch/operations/loop.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/loop.py:19]: Sounds good, I'll be able to fix this in December / January.

File: onnx2pytorch/operations/loop.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/loop.py:26]: These delayed imports are to avoid Python import errors due to circular import from a partially-initialized module. `operations.convert_operations` needs to import `Loop`, but `Loop` also needs to import `operations.convert_operations` to convert the operations inside the subgraph for the body of the loop. Meanwhile, `ConvertModel.__init__` needs `operations.convert_operations`, while `ConvertModel.forward` needs `Loop`.

File: onnx2pytorch/convert/model.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/model.py:63]: Agreed, sounds good. I'll be able to take care of the issue in December/January.

File: onnx2pytorch/convert/layer.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/layer.py:14]: I tried removing the `np.copy` and it turns out our training code using onnx2pytorch still works fine -- I'm guessing there was a bug on our end that's since been fixed. Plus I added a new test `test_train.py` to verify that `np.copy` is unnecessary.

File: onnx2pytorch/operations/pad.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/pad.py]: Just some debugging cruft which snuck through -- now removed.

File: onnx2pytorch/operations/loop.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/loop.py:71]: Same answer as above (circular import issue).

File: onnx2pytorch/convert/model.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/model.py]: np.copy not needed, so I've reverted my changes

File: onnx2pytorch/convert/layer.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/convert/layer.py]: np.copy not needed, so I've reverted my changes

File: onnx2pytorch/operations/constantofshape.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/constantofshape.py]: Done! And added tests in `test_constantofshape.py`

File: onnx2pytorch/operations/gathernd.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/gathernd.py]: Done! Interestingly enough, in torch array access is a bit janky, but fortunately `torch.split` fixes it:
```
X = torch.rand(2, 2) # assume returns tensor([[a, b], [c, d]])
X[[[0, 1], [0, 1]]]  # returns tensor([a, d])
X[torch.tensor([[0, 1], [0, 1]])]  # returns tensor([[[a, b], [c, d]], [[a, b], [c, d]]])
X[torch.split(torch.tensor([[0, 1], [0, 1]]), 1, 0)]  # returns tensor([a, d])
```

File: onnx2pytorch/operations/add.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/add.py]: I've reverted this change back to `all`. Somehow our code using onnx2pytorch now works with `all`, and I can't remember my reasoning behind the change...

File: onnx2pytorch/convert/operations.py
Comment (Talmaj) [inline onnx2pytorch/convert/operations.py:265]: Tile is not imported in `onnx2pytorch/operations/__init__.py`

File: onnx2pytorch/operations/__init__.py
Comment (Talmaj) [inline onnx2pytorch/operations/__init__.py:32]: Import Tile.

File: onnx2pytorch/operations/loop.py
Comment (Talmaj) [reply onnx2pytorch/operations/loop.py:26]: What about doing here: 
```
self.ops = import_module("onnx2pytorch.convert.operations")
self.c = import_module("onnx2pytorch.constants")
```
And then use it in forward method.

In the beginning of the file you should add: `from importlib import import_module`

File: onnx2pytorch/operations/slice.py
Comment (Talmaj) [inline onnx2pytorch/operations/slice.py:6]: Can you please add a short docstring what this function does and in what case it is used?

File: onnx2pytorch/operations/split.py
Comment (Talmaj) [inline onnx2pytorch/operations/split.py:12]: Put a default value for enable_pruning=False

File: onnx2pytorch/operations/unsqueeze.py
Comment (Talmaj) [inline onnx2pytorch/operations/unsqueeze.py:23]: Why the change? Was it wrong before?

File: setup.py
Comment (Talmaj) [inline setup.py:20]: Why is this requirement necessary?

**Review by Talmaj — COMMENTED** [review]
I did not check all the tests in detail. Did you base them on the examples from onnx operators?

File: setup.py
Comment (calvinmccarter-at-lightmatter) [reply setup.py:20]: It's needed for the NonMaxSuppression operation.

File: onnx2pytorch/operations/unsqueeze.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/unsqueeze.py:23]: Yes, prior to the change, it was failing `test_unsqueeze_unsorted_axes` (which is taken from the same example in the ONNX operator readme).

**Comment by calvinmccarter-at-lightmatter** [comment]
> I did not check all the tests in detail. Did you base them on the examples from onnx operators?

Yeah, they're all just taken from the onnx operators examples.

File: onnx2pytorch/operations/slice.py
Comment (calvinmccarter-at-lightmatter) [reply onnx2pytorch/operations/slice.py:6]: Added a docstring -- Pytorch still doesn't support slicing tensors with a negative step (gives "step must be greater than zero").

File: onnx2pytorch/convert/model.py
Comment (Talmaj) [inline onnx2pytorch/convert/model.py:87]: Let's set enable_pruning=False by default.

**Review by Talmaj — CHANGES_REQUESTED** [review]
There are few last things that need to be addressed before merging:
- [x] add a test for Reshape operator with enable_pruning=False
- [x] add a test for Split operator with enable_pruning=False
- [x] add a test for _to_positive_step in slice.py
- [ ] add a test for get_ops_names in utils.py
- [ ] move test_loop.py from onnx2pytorch/convert to onnx2pytorch/operations

I would also like to improve the test coverage with each merge. Currently cast and instancenorm operations have the lowest test [coverage](https://38-252023841-gh.circle-artifacts.com/0/htmlcov/index.html). Please add:
- [ ] test for cast
- [ ] test for instancenorm

**Review by Talmaj — APPROVED** [review]

**Comment by Talmaj** [comment]
@calvinmccarter-at-lightmatter do you know any pre-trained LSTM model that is from now on supported?

**Comment by calvinmccarter-at-lightmatter** [comment]
I've attached the prediction portion of the MLCommons Inference RNN-T model. The original PyTorch checkpoint of the entire model is here: https://doi.org/10.5281/zenodo.3662521. I then exported the prediction submodel to ONNX.
[rnnt_prediction.zip](https://github.com/ToriML/onnx2pytorch/files/7306093/rnnt_prediction.zip) Hope this helps!

**Comment by calvinmccarter-at-lightmatter** [comment]
This repo has LSTM and GRU models: https://github.com/IBM/ai-on-z-fraud-detection . I'll verify that these are converted correctly in the near future.

**Pull request merged**
